pkg/build/controller/buildpod/factory.go

```diff
@@ -0,0 +1,213 @@
+package buildpod
+
+import (
+	"errors"
+	"time"
+
+	"github.com/golang/glog"
+	kapi "k8s.io/kubernetes/pkg/api"
+	kerrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/client/cache"
+	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	kcontroller "k8s.io/kubernetes/pkg/controller"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	buildclient "github.com/openshift/origin/pkg/build/client"
+	"github.com/openshift/origin/pkg/build/controller/policy"
+	strategy "github.com/openshift/origin/pkg/build/controller/strategy"
+	buildutil "github.com/openshift/origin/pkg/build/util"
+	osclient "github.com/openshift/origin/pkg/client"
+)
+
+const (
+	// We must avoid processing build pods until the build and pod stores have synced.
+	// If they haven't synced, to avoid a hot loop, we'll wait this long between checks.
+	storeSyncedPollPeriod = 100 * time.Millisecond
+	maxRetries            = 60
+)
+
+// NewBuildPodController creates a new BuildPodController.
+func NewBuildPodController(buildInformer, podInformer cache.SharedIndexInformer, kc kclientset.Interface, oc osclient.Interface) *BuildPodController {
+	buildUpdater := buildclient.NewOSClientBuildClient(oc)
+	buildLister := buildclient.NewOSClientBuildClient(oc)
+	c := &BuildPodController{
+		buildUpdater: buildUpdater,
+		secretClient: kc.Core(),
+		queue:        workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+
+	c.runPolicies = policy.GetAllRunPolicies(buildLister, buildUpdater)
+
+	c.buildStore.Indexer = buildInformer.GetIndexer()
+	c.podStore.Indexer = podInformer.GetIndexer()
+	podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: c.updatePod,
+		DeleteFunc: c.deletePod,
+	})
+
+	c.buildStoreSynced = buildInformer.HasSynced
+	c.podStoreSynced = podInformer.HasSynced
+
+	return c
+}
+
+// Run begins watching and syncing.
+func (c *BuildPodController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	// Wait for the build store to sync before starting any work in this controller.
+	ready := make(chan struct{})
+	go c.waitForSyncedStores(ready, stopCh)
+	select {
+	case <-ready:
+	case <-stopCh:
+		return
+	}
+
+	go wait.Until(c.worker, time.Second, stopCh)
+	<-stopCh
+	glog.Infof("Shutting down build pod controller")
+	c.queue.ShutDown()
```

Comment:
@kargakis it doesn't get deferred in any of the deployment controllers... 
https://github.com/openshift/origin/blob/master/pkg/deploy/controller/deployment/factory.go#L83
https://github.com/openshift/origin/blob/master/pkg/deploy/controller/deploymentconfig/factory.go#L91
https://github.com/openshift/origin/blob/master/pkg/deploy/controller/generictrigger/factory.go#L76

what would be the reason for doing it here?